atc/creds/vault/api_client.go

```diff
@@ -196,6 +200,73 @@ func (ac *APIClient) baseClient() (*vaultapi.Client, error) {
 	return client, nil
 }
 
+func (ac *APIClient) configureTLS(config *tls.Config) error {
+	if ac.tlsConfig.CACert != "" || ac.tlsConfig.CACertFile != "" || ac.tlsConfig.CAPath != "" {
+		rootConfig := &rootcerts.Config{
+			CAFile:        ac.tlsConfig.CACertFile,
+			CAPath:        ac.tlsConfig.CAPath,
+			CACertificate: []byte(ac.tlsConfig.CACert),
+		}
+
+		if err := rootcerts.ConfigureTLS(config, rootConfig); err != nil {
+			return err
+		}
+	}
+
+	if ac.tlsConfig.ClientCertFile != "" {
+		content, err := ioutil.ReadFile(ac.tlsConfig.ClientCertFile)
+		if err != nil {
+			return err
+		}
+
+		ac.tlsConfig.ClientCert = string(content)
+	}
+
+	if ac.tlsConfig.ClientKeyFile != "" {
+		content, err := ioutil.ReadFile(ac.tlsConfig.ClientKeyFile)
+		if err != nil {
+			return err
+		}
+
+		ac.tlsConfig.ClientKey = string(content)
+	}
+
+	if ac.tlsConfig.Insecure {
+		config.InsecureSkipVerify = true
+	}
+
+	var clientCert tls.Certificate
+	foundClientCert := false
+
+	switch {
+	case ac.tlsConfig.ClientCert != "" && ac.tlsConfig.ClientKey != "":
+		var err error
+		clientCert, err = tls.X509KeyPair([]byte(ac.tlsConfig.ClientCert), []byte(ac.tlsConfig.ClientKey))
+		if err != nil {
+			return err
+		}
+
+		foundClientCert = true
+	case ac.tlsConfig.ClientCert != "" || ac.tlsConfig.ClientKey != "":
+		return fmt.Errorf("both client cert and client key must be provided")
+	}
+
+	if foundClientCert {
+		// We use this function to ignore the server's preferential list of
+		// CAs, otherwise any CA used for the cert auth backend must be in the
+		// server's CA pool
+		config.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			return &clientCert, nil
+		}
+	}
+
+	if ac.tlsConfig.ServerName != "" {
+		config.ServerName = ac.tlsConfig.ServerName
+	}
+
+	return nil
```

Comment:
whoa, deja vu https://github.com/hashicorp/vault/blob/2c21ef4df684d60030f6282997a0ad38f824bcaf/api/client.go#L190-L232. This part seemed like too much code with too few tests... and I bet the hashicorp codebase has tests for all these branches. Would it make sense to write the `tlsConfig.ClientCert` to a temporary file or something as a way to actually properly reuse their code?

This intuitively feels like the kind of place an unsuspecting person would just inject a little change without realizing that it's supposed to pretty much identically simulate the equivalent function in that `vault/api` package...

Reply:
The tests actually cover most of it, except for `insecure_skip_verify`, plus a few more 'sad path' cases could probably be added to ensure the values are respected. I hand-validated it by changing the values in the tests and watching them break but didn't commit to adding additional examples for them. I can go back and add them.

To be honest, this is all pretty basic TLS configuration that is either correct or totally broken; there's not much to simulate in `vault/api` that wouldn't be done in the exact same way in other projects or if we had just written it ourselves. It was just disappointing that I had to carry over so much just because their client wrapper doesn't support configuring via cert content instead of file paths everywhere. :cry:

Reply:
yeah, i'll just update our internal wiki with certstrap instructions